### PR TITLE
Run single test shortcut + stefanforsberg f-test-result PR

### PR DESCRIPTION
**Pull request by janaka**
**tldr; This PR implements shortcuts and includes PR #32** 

This PR is the first part of proposed changes to include some of the productivity features found in Visual Studio such as keyboard shortcuts and working on filtered lists of tests which are important when working in a code base with a large number of tests. 

My contribution in this PR is making executing a single selected unit test via a shortcut possible.

Later I noticed Stefan Forsberg's PR #32 which adds test results to the tree view. Something I was planning on adding as devs where I work had mentioned as a shortcoming. I didn't realise there was a PR until after I did a little work to merge into my branch. There were a couple of conflicts and I believe I've improved the code to be a bit more generic where there was conflict. Therefore I'm proposing the combined PR (I realise this isn't best practice with unrelated changes so if you have an alternative approach please let me know).    

## Other Changes

- Add: .NET Core Project with tests for testing. Contains xUnit. Should be able to add MSTest and NUnit. Probably help with the multi-test project (#18) and *.sln file support (#31) requests. 
- Fix: Single test execution. Was running all tests.  (related to #26) 
- Add: `onSelectedUnitTest()` "event" for firing logic with the context of the selected TestNode.
- Add: ability to apply context menus based on the type of node using the [`when` clause context](https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts) for `view/item/context` contributions. Only implements UnitTest type nodes in this PR. 
- Add: `category` to all command contributions such that they are grouped in the command palette view.
- Refactor: deprecated variable ${workspaceRoot} replace with ${workspaceFolder}
- Change: I think PR #32 meant clicking on the test in the tree would focus the code on the test. This PR takes that out in favour of a right-click context menu item (shortcut can easily be added). If we want to this back it's a 1 line addition to the `onSelectedUnitTest()` event.

## ToDo
- [x] Test with xUnit (MacOS)
- [x] Test with xUnit (Windows)
- [ ] Test with MSTest (MacOS)
- [ ] Test with MSTest (Windows)
- [ ] Test with NUnit (MacOS)
- [ ] Test with NUnit (Windows)
 

## Broader Motivation
@formulahendry Thanks for your work getting this project kicked off. Obviously, I'd like to try to contribute and help move it forward. 

For a bit of context, the company I work for, Hudl, is moving our entire stack from .NET on Windows over to .NET Core on Docker and Linux. We'll be supporting "native" local development on Windows and MacOS (as opposed to having to use Parallels). VS Code is our goto IDE at present for MacOS (and Linux), however, the lack of certain features, such as test management, compared to Visual Studio are blockers. After all the entire reason for investing in a move to .NET Core and Docker is to enable our product development to be more efficient and productive.

### CODE REVIEW

**Comment by stefanforsberg** [comment]
Tested and confirmed working with xUnit (Windows)

**Comment by janaka** [comment]
Hey @formulahendry, does this look good?

**Comment by formulahendry** [comment]
Thanks @janaka and @stefanforsberg a lot! Could you please also help test it with MSTest and NUnit? 😄 If the color feature only works for xUnit, I just want to make sure there is no regression on MSTest and NUnit.

**Comment by stefanforsberg** [comment]
@formulahendry I can help with mstest and nunit on windows.

Regarding the color feature I wanted your opinion on nunit/mstest. The output from dotnet test -t is given in the form Name.Space.Class.Method for xunit but only Method for nunit (and mstest maybe?). So when we try to parse the result from the run tests we can only match by the method name which is complicated in the scenario where you have two test methods _with the same name_. Now, if one of these methods pass and another fail we can't mark the correct one in the tree view.

Therefor I opted for only displaying the color coding when we can be sure that the correct test is marked with the correct result.

Any thoughts on this?

**Comment by janaka** [comment]
I'm first going to work on getting MSTest and NUnit unit tests into the test project ( 😁). This should make testing easier now and in the future. If either of you know how to write unit test to test the various test runners. I have no experience unit testing VS Coe extension or mocha etc.

**Comment by stefanforsberg** [comment]
I can work on that but this PR already contain so much different stuff that I'm hesitant to add more to it tbh.

**Comment by formulahendry** [comment]
> Therefor I opted for only displaying the color coding when we can be sure that the correct test is marked with the correct result.

Agree. If we could only match xUnit with the correct result, then we could enable color feature for xUnit only.

> As far as I can tell this fix to run one test works for xUnit but breaks the functionality for mstest

Thanks @stefanforsberg for testing. Do you mean running single mstest would not work? 
I am ok with color feature only works on xUnit. We just need to make sure all the existing feature are still working. I have create a test plan to cover main features: https://github.com/formulahendry/vscode-dotnet-test-explorer/wiki/Test-Plan  Feel free to help update it. 😃 

> I'm first going to work on getting MSTest and NUnit unit tests into the test project

@janaka Cool. Thank you for improving the test coverage! BTW, I think it would be a good idea if we could separate PR (make PR not too large) for each feature in the future. 😄

**Comment by janaka** [comment]
Agree with both of you regarding the PR size, current size isn't ideal.

**Comment by stefanforsberg** [comment]
@formulahendry @janaka Ideally we would have an automated test suite that would at least test the basic stuff that is in the test plan. I don't have any experience writing these type of tests but it should be doable I think. I am going skiing for a week now so I won't be able to do any work on this until next weekend 👼 

I'm not sure what's the best way forward, to fix the various issues in this PR or do more small PRs (just mine that added the color coding also adds setting for where to store test result files as well as navigate to test code. These should maybe be three different pull request I think).

**Comment by formulahendry** [comment]
I suggest making each feature into small PRs. It's good for code review.

**Comment by cannontrodder** [comment]
I can test this with mstest on both Windows and OSX.  At the moment it seemed to me that testing a single test failed but notes above seem to suggest that does not work? What I do see working is the pathForResultFile setting from @stefanforsberg which gives me codelens icons on passing tests.
